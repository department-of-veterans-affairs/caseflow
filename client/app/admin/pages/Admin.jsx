
export const Admin = ({
  // sys_admin,
  // errors,
  // hearing,
  // showDivider,
  // appellantTitle,
  // formFieldsOnly,
  // setIsValidEmail,
  // update,
  // actionType
}) => {
  return (
    <h1>System Admin UI</h1>
  );
};