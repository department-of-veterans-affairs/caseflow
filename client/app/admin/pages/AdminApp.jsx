import React from 'react';
import { BrowserRouter, Switch } from 'react-router-dom';
import PropTypes from 'prop-types';
import PageRoute from '../../components/PageRoute';
import NavigationBar from '../../components/NavigationBar';
import { LOGO_COLORS } from '../../constants/AppConstants';
import AppFrame from '../../components/AppFrame';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import Footer from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/Footer';
import GenerateButton from '../components/GenerateButton';
import { sendExtractRequest } from '../actions';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

class AdminApp extends React.PureComponent {
  render = () => <BrowserRouter basename="/admin">
    <NavigationBar
      wideApp
      defaultUrl="/admin"
      userDisplayName={this.props.userDisplayName}
      dropdownUrls={this.props.dropdownUrls}
      applicationUrls={this.props.applicationUrls}
      logoProps={{
        overlapColor: LOGO_COLORS.QUEUE.OVERLAP,
        accentColor: LOGO_COLORS.QUEUE.ACCENT,
      }}
      appName="System Admin"
    >
      <AppFrame wideApp>
        <AppSegment filledBackground>
          <h1>System Admin UI</h1>
          <div />
          <div>
            <Switch>
              <PageRoute
                exact
                path="/admin"
                title="admin"
                render={this.admin}
              />
              <GenerateButton
                sendExtractRequest={this.props.sendExtractRequest}
                {...this.props.extractReducer}
              />
            </Switch>
          </div>
        </AppSegment>
      </AppFrame>
    </NavigationBar>
    <Footer
      wideApp
      appName=""
      feedbackUrl={this.props.feedbackUrl}
      buildDate={this.props.buildDate}
    />
  </BrowserRouter>
}

AdminApp.propTypes = {
  userDisplayName: PropTypes.string.isRequired,
  dropdownUrls: PropTypes.array,
  applicationUrls: PropTypes.array,
  feedbackUrl: PropTypes.string.isRequired,
  buildDate: PropTypes.string,
  sendExtractRequest: PropTypes.func,
};

const mapStateToProps = (state) => {
  return {
    ...state
  };
};

const mapDispatchToProps = (dispatch) => bindActionCreators({
  sendExtractRequest
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AdminApp);
