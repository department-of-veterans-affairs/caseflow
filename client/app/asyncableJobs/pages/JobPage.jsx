/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import AsyncModelNav from '../components/AsyncModelNav';
import JobRestartButton from '../components/JobRestartButton';
import JobNotes from '../components/JobNotes';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

class AsyncableJobPage extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      restarted: 0
    };
  }

  formatDate = (datetime) => {
    if (!datetime) {
      return 'n/a';
    }

    return moment(datetime).format(DATE_TIME_FORMAT);
  }

  render = () => {
    const { job, notes } = this.props;

    return <div className="cf-asyncable-job-table">
      <h1>{this.props.asyncableJobKlass} Job {job.id}</h1>
      <AsyncModelNav models={[]} fetchedAt={this.props.fetchedAt} />
      <table className="cf-job-details">
        <tbody>
          <tr>
            <th>ID</th>
            <td>{job.id}</td>
          </tr>
          <tr>
            <th>Veteran</th>
            <td>{job.veteran_file_number}</td>
          </tr>
          <tr>
            <th>Originally Submitted</th>
            <td>{this.formatDate(job.submitted_at)}</td>
          </tr>
          <tr>
            <th>Last Submitted</th>
            <td>{this.formatDate(job.last_submitted_at)}</td>
          </tr>
          <tr>
            <th>Last Attempted</th>
            <td>{this.formatDate(job.attempted_at)}</td>
          </tr>
          <tr>
            <th>Canceled</th>
            <td>{this.formatDate(job.canceled_at)}</td>
          </tr>
          <tr>
            <th>Processed</th>
            <td>{this.formatDate(job.processed_at)}</td>
          </tr>
          <tr>
            <th>Error</th>
            <td>{job.error}</td>
          </tr>
          <tr>
            <th>User</th>
            <td><a href={`/intake/manager?user_css_id=${job.user}`}>{job.user}</a></td>
          </tr>
        </tbody>
      </table>
      <div>
        <JobRestartButton job={job} page={this} />
      </div>
      <div>
        <JobNotes job={job} notes={notes} />
      </div>
    </div>;
  }
}

const JobPage = connect(
  (state) => ({
    job: state.job,
    notes: state.notes,
    fetchedAt: state.fetchedAt,
    models: state.models,
    asyncableJobKlass: state.asyncableJobKlass
  })
)(AsyncableJobPage);

export default JobPage;
