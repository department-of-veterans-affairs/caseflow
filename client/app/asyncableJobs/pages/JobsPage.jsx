import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import Button from '../../components/Button';
import Table from '../../components/Table';

import ApiUtil from '../../util/ApiUtil';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

class AsyncableJobsPage extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      jobsRestarted: {},
      jobsRestarting: {}
    };
  }

  restartJob = (job) => {
    let jobsRestarting = { ...this.state.jobsRestarting };

    jobsRestarting[job.id] = true;
    this.setState({ jobsRestarting });
    this.sendRestart(job);
  }

  sendRestart = (job) => {
    let page = this;

    ApiUtil.patch(`/asyncable_jobs/${job.klass}/jobs/${job.id}`, {}).
      then(
        (response) => {
          const responseObject = JSON.parse(response.text);

          Object.assign(job, responseObject);

          // TODO null it on server? responseObject.error;
          job.error = '';

          job.restarted = true;

          let jobsRestarted = { ...page.state.jobsRestarted };
          let jobsRestarting = { ...page.state.jobsRestarting };

          jobsRestarted[job.id] = true;
          jobsRestarting[job.id] = false;
          page.setState({
            jobsRestarted,
            jobsRestarting
          });
        },
        (error) => {
          throw error;
        }
      ).
      catch((error) => error);
  }

  getButtonClassNames = (job) => {
    let classNames = ['usa-button'];

    if (this.state.jobsRestarting[job.id] || this.state.jobsRestarted[job.id]) {
      classNames.push('usa-button-disabled');
    }

    return classNames;
  }

  getButtonText = (job) => {
    let txt = 'Restart';

    if (this.state.jobsRestarting[job.id]) {
      txt = 'Restarting';
    } else if (this.state.jobsRestarted[job.id]) {
      txt = 'Restarted';
    }

    return txt;
  }

  render = () => {
    const rowObjects = this.props.jobs;

    if (rowObjects.length === 0) {
      return 'Success! There are no pending jobs.';
    }

    const columns = [
      {
        header: 'Name',
        valueFunction: (job) => {
          return <a href={`/asyncable_jobs/${job.klass}/jobs`}>{job.klass}</a>;
        }
      },
      {
        header: 'Submitted',
        valueFunction: (job) => {
          if (job.submitted_at === 'restarted') {
            return job.submitted_at;
          }

          return moment(job.submitted_at).format(DATE_TIME_FORMAT);
        }
      },
      {
        header: 'Last Attempted',
        valueFunction: (job) => {
          if (!job.attempted_at) {
            return 'never';
          }

          return moment(job.attempted_at).format(DATE_TIME_FORMAT);
        }
      },
      {
        header: 'Error',
        valueFunction: (job) => {
          return <span className="cf-job-error">{job.error}</span>;
        }
      },
      {
        header: 'Veteran',
        valueName: 'veteran_file_number'
      },
      {
        header: 'Restart',
        align: 'right',
        valueFunction: (job) => {
          return <Button
            id={`job-${job.klass}-${job.id}`}
            loading={this.state.jobsRestarting[job.id]}
            loadingText="Restarting..."
            onClick={() => {
              this.restartJob(job);
            }}
            classNames={this.getButtonClassNames(job)}
          >{this.getButtonText(job)}</Button>;
        }
      }
    ];

    const rowClassNames = (rowObject) => {
      return rowObject.restarted ? 'cf-success' : '';
    };

    return <div className="cf-asyncable-jobs-table">
      <h1>{this.props.asyncableJobKlass} Jobs</h1>
      <div>
        <strong>Last updated:</strong> {moment(this.props.fetchedAt).format(DATE_TIME_FORMAT)}
        &nbsp;&#183;&nbsp;
        <a href="/jobs">All jobs</a>
      </div>
      <Table columns={columns} rowObjects={rowObjects} rowClassNames={rowClassNames} slowReRendersAreOk />
    </div>;
  }
}

const JobsPage = connect(
  (state) => ({
    jobs: state.jobs,
    fetchedAt: state.fetchedAt,
    asyncableJobKlass: state.asyncableJobKlass
  })
)(AsyncableJobsPage);

export default JobsPage;
