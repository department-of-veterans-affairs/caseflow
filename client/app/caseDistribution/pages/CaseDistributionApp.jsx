import React from 'react';
import PropTypes from 'prop-types';
import CaseDistributionContent from '../components/CaseDistributionContent';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  loadLevers,
  loadHistory,
  setUserIsAcdAdmin
} from '../reducers/levers/leversActions';

class CaseDistributionApp extends React.PureComponent {
  constructor(props) {
    super(props);
    this.props.loadLevers(this.props.acdLeversForStore);
    this.props.loadHistory(this.props.acd_history);
    this.props.setUserIsAcdAdmin(this.props.user_is_an_acd_admin);
  }

  render() {
    return (
      <div>
        <div>
          <CaseDistributionContent />
        </div>
      </div>
    );

  }
}

CaseDistributionApp.propTypes = {
  acdLeversForStore: PropTypes.object,
  acd_history: PropTypes.array,
  user_is_an_acd_admin: PropTypes.bool,
  loadLevers: PropTypes.func,
  loadHistory: PropTypes.func,
  setUserIsAcdAdmin: PropTypes.func,
};

// eslint-disable-next-line no-unused-vars
const mapStateToProps = (state) => ({
});

const mapDispatchToProps = (dispatch) => (
  bindActionCreators({
    loadLevers,
    loadHistory,
    setUserIsAcdAdmin
  }, dispatch)
);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CaseDistributionApp);
