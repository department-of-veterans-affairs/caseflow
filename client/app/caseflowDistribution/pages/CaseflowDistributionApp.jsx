import React from 'react';
import PropTypes from 'prop-types';
import CaseflowDistributionContent from '../components/CaseflowDistributionContent';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  initialLoad
} from '../reducers/Levers/leversActions';

class CaseflowDistributionApp extends React.PureComponent {
  constructor(props) {
    super(props);
    this.props.initialLoad(this.props.acdLeversForStore);
  }

  render() {
    return (
      <div>
        <div> {/* Wrapper*/}
          <CaseflowDistributionContent
            levers = {this.props.acd_levers}
            saveChanges = {() => {}}
            formattedHistory={this.props.acd_history}
            leverStore={this.props.leverStore}
            isAdmin = {this.props.user_is_an_acd_admin}
            sectionTitles = {this.props.sectionTitles}
          />
        </div>
      </div>
    );

  }
}

CaseflowDistributionApp.propTypes = {
  acd_levers: PropTypes.object,
  acd_history: PropTypes.array,
  user_is_an_acd_admin: PropTypes.bool,
  leverStore: PropTypes.any,
  sectionTitles: PropTypes.array,
  initialLoad: PropTypes.func,
  acdLeversForStore: PropTypes.object
};

// eslint-disable-next-line no-unused-vars
const mapStateToProps = (state) => ({
});

const mapDispatchToProps = (dispatch) => (
  bindActionCreators({
    initialLoad
  }, dispatch)
);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CaseflowDistributionApp);
