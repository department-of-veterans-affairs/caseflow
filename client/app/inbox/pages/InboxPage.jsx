/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import Button from '../../components/Button';
import Table from '../../components/Table';
import EasyPagination from '../../components/EasyPagination';

import ApiUtil from '../../util/ApiUtil';

const DATE_TIME_FORMAT = 'ddd MMM DD YYYY [at] HH:mm';

class InboxMessagesPage extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      markedRead: {}
    };
  }

  markMessageRead = (msg) => {
    const markedRead = { ...this.state.markedRead };

    markedRead[msg.id] = true;
    this.setState({ markedRead });
    this.sendMessageRead(msg);
  }

  sendMessageRead = (msg) => {
    const page = this;

    ApiUtil.patch(`/inbox/messages/${msg.id}`, { data: { message_action: 'read' } }).
      then(
        (response) => {
          const responseObject = JSON.parse(response.text);

          Object.assign(msg, responseObject);

          const markedRead = { ...page.state.markedRead };

          markedRead[msg.id] = true;
          page.setState({
            markedRead
          });
        },
        (error) => {
          throw error;
        }
      ).
      catch((error) => error);
  }

  getButtonText = (msg) => {
    let txt = 'Mark as read';

    if (msg.read_at) {
      txt = `Read ${this.formatDate(msg.read_at)}`;
    }

    return txt;
  }

  formatDate = (datetime) => {
    return moment(datetime).format(DATE_TIME_FORMAT);
  }

  markAsReadButtonDisabled = (msg) => {
    if (this.state.markedRead[msg.id] || msg.read_at) {
      return true;
    }

    return false;
  }

  render = () => {
    const rowObjects = this.props.messages;

    if (rowObjects.length === 0) {
      return <div>
        <h1>Success! You have no unread messages.</h1>
      </div>;
    }

    const columns = [
      {
        header: 'Received',
        valueFunction: (msg) => {
          return this.formatDate(msg.created_at);
        }
      },
      {
        header: 'Message',
        valueFunction: (msg) => {
          // allow raw html since we control message content.
          return <span className="cf-inbox-message" dangerouslySetInnerHTML={{ __html: msg.text }} />;
        }
      },
      {
        align: 'right',
        valueFunction: (msg) => {
          return <Button
            id={`inbox-message-${msg.id}`}
            title={`message ${msg.id}`}
            disabled={this.markAsReadButtonDisabled(msg)}
            onClick={() => {
              this.markMessageRead(msg);
            }}
          >{this.getButtonText(msg)}</Button>;
        }
      }
    ];

    const rowClassNames = (msg) => {
      if (this.state.markedRead[msg.id] || msg.read_at) {
        return 'cf-inbox-message-read';
      }

      return 'cf-inbox-message';
    };

    return <div className="cf-inbox-table">
      <h1>Inbox</h1>
      <hr />
      <div>
        Messages will remain in the intake box for 120 days. After such time, messages will be removed.
      </div>
      <Table columns={columns} rowObjects={rowObjects} rowClassNames={rowClassNames} slowReRendersAreOk />
      <EasyPagination currentCases={rowObjects.length} pagination={this.props.pagination} />
    </div>;
  }
}

const InboxPage = connect(
  (state) => ({
    messages: state.messages,
    pagination: state.pagination
  })
)(InboxMessagesPage);

export default InboxPage;
