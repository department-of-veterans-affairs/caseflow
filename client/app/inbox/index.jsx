/* eslint-disable react/prop-types */

import React from 'react';
import ReduxBase from '../components/ReduxBase';
import NavigationBar from '../components/NavigationBar';
import IntakeLink from '../components/IntakeLink';
import { BrowserRouter } from 'react-router-dom';
import PageRoute from '../components/PageRoute';
import AppFrame from '../components/AppFrame';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import { LOGO_COLORS } from '../constants/AppConstants';
import Footer from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/Footer';

import InboxPage from './pages/InboxPage';
import { inboxReducer, mapDataToInitialState } from './reducers';

class Inbox extends React.PureComponent {
  render() {
    const Router = this.props.router || BrowserRouter;
    const initialState = mapDataToInitialState(this.props);
    const appName = 'Inbox';

    return <ReduxBase initialState={initialState} reducer={inboxReducer}>
      <Router {...this.props.routerTestProps}>
        <div>
          <NavigationBar
            appName={appName}
            logoProps={{
              accentColor: LOGO_COLORS.INTAKE.ACCENT,
              overlapColor: LOGO_COLORS.INTAKE.OVERLAP
            }}
            rightNavElement={<IntakeLink />}
            userDisplayName={this.props.userDisplayName}
            dropdownUrls={this.props.dropdownUrls}
            topMessage={null}
            defaultUrl="/">
            <AppFrame>
              <AppSegment filledBackground>
                <div>
                  <PageRoute
                    exact
                    path="/inbox"
                    title="Inbox | Caseflow"
                    component={InboxPage} />
                </div>
              </AppSegment>
            </AppFrame>
          </NavigationBar>
          <Footer
            appName={appName}
            feedbackUrl={this.props.feedbackUrl}
            buildDate={this.props.buildDate}
          />
        </div>
      </Router>
    </ReduxBase>;
  }
}

export default Inbox;
