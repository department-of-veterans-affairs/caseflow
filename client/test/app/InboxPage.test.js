import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe } from 'jest-axe';

import { InboxMessagesPage } from '../../app/inbox/pages/InboxPage';
import { emptyMessages, allUnreadMessages, oneReadAndOneUnreadMessages } from '../data/inbox';

const defaultProps = {
  messages: emptyMessages,
  pagination: {
    current_page: 1,
    page_size: 50,
    total_items: 2,
    total_pages: 1
  }
};

const paginationProps = defaultProps.pagination;

const successMessage = 'Success! You have no unread messages.';
const messagesRemovedMessage =
  'Messages will remain in the intake box for 120 days. After such time, messages will be removed.';
const paginationOptions =
  `Viewing ${paginationProps.current_page}-${paginationProps.total_items} of ${paginationProps.total_items} total`;

const setupComponent = (props = {}) => {
  return render(
    <InboxMessagesPage {...defaultProps}{...props} />
  );
};

describe('InboxPage rendering with an empty inbox', () => {
  it('renders correctly', async () => {
    const { container } = setupComponent();

    expect(container).toMatchSnapshot();
  });

  it('passes a11y testing', async () => {
    const { container } = setupComponent();

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('renders a the success message', () => {
    setupComponent();

    expect(screen.getByText(successMessage)).toBeInTheDocument();
  });
});

describe('InboxPage rendering with data', () => {
  const setupMessages = (messages) => {
    defaultProps.messages = messages;
    setupComponent();
  };

  it('renders no success message', () => {
    setupMessages(allUnreadMessages);

    expect(screen.queryByText(successMessage)).not.toBeInTheDocument();
  });

  it('has a message about when the messages are removed', () => {
    setupMessages(allUnreadMessages);

    expect(screen.getByText(messagesRemovedMessage)).toBeInTheDocument();
  });

  it('renders the correct pagination options', () => {
    setupMessages(allUnreadMessages);

    expect(screen.getByText(paginationOptions)).toBeInTheDocument();
  });

  it('renders an inbox with two unread messages', () => {
    setupMessages(allUnreadMessages);

    const trElements = screen.getAllByRole('row');

    expect(trElements.length - 1).toBe(2);

    const unreadButtons = screen.getAllByRole('button');

    expect(unreadButtons.length).toBe(2);
    for (let button of unreadButtons) {
      expect(button).toBeEnabled();
    }
  });

  it('renders an inbox with one read and one unread messages', () => {
    setupMessages(oneReadAndOneUnreadMessages);

    const trElements = screen.getAllByRole('row');

    expect(trElements.length - 1).toBe(2);

    const allButtons = screen.getAllByRole('button');

    expect(allButtons[0]).toBeInTheDocument();
    expect(allButtons[0]).not.toBeEnabled();
  });
});
