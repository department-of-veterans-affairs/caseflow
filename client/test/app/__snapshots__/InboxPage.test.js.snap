// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InboxPage rendering with an empty inbox renders correctly 1`] = `
<div>
  <div
    class="cf-txt-c"
  >
    <h1>
      Success! You have no unread messages.
    </h1>
  </div>
</div>
`;
